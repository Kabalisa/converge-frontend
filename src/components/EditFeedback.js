import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Input } from './commons/Input';
import Select from './commons/SelectInput';
import MrmModal from './commons/Modal';
import ActionButtons from './commons/ActionButtons';
import '../assets/styles/editfeedback.scss';

class EditFeedback extends Component {
  state = {
    question: this.props.question,
    type: this.props.type,
    startDate: this.props.startDate,
    duration: this.props.duration,
    closeModal: false,
  };

  handleCloseModal = () => {
    this.setState({ closeModal: true });
  }

  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  }

  handleInputChange = ({ target: { name, value } }) => {
    this.setState({ [name]: value });
  }

  handleEditFeedback = (event) => {
    event.preventDefault();
    /** submit logic here */
    this.handleCloseModal();
  }

  render() {
    const {
      question, type, startDate, duration, closeModal,
    } = this.state;

    return (
      <MrmModal
        title="Edit Question"
        buttonText="Edit"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="edit-feedback-modal"
        modalButtonClassName="edit-button"
      >
        <form className="modal-form" onSubmit={this.handleEditFeedback}>
          <Input
            labelName="Question"
            labelClass="label1"
            inputClass="input1"
            name="question"
            id="question"
            value={question}
            onChange={this.handleInputChange}
          />
          <Select
            labelText="Type"
            name="type"
            id="type"
            value={type}
            onChange={this.handleInputChange}
            options={[]}
            wrapperClassName="label1"
            placeholder="Rate"
            selectInputClassName="edit-feedback-select"
          />
          <Input
            labelName="Start Date"
            labelClass="label1"
            inputClass="input1"
            name="startDate"
            id="startDate"
            value={startDate}
            placeholder="03 Nov 2018"
            onChange={this.handleInputChange}
          />
          <Input
            labelName="Duration"
            labelClass="label1"
            inputClass="input1"
            name="duration"
            id="duration"
            value={duration}
            placeholder="3 Weeks"
            onChange={this.handleInputChange}
          />
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            actionButtonText="Save Changes"
          />
        </form>
      </MrmModal>
    );
  }
}

EditFeedback.propTypes = {
  question: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  startDate: PropTypes.string.isRequired,
  duration: PropTypes.string.isRequired,
};

export default EditFeedback;

