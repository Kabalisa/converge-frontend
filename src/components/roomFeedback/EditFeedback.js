import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Input } from '../commons/Input';
import Select from '../commons/SelectInput';
import MrmModal from '../commons/Modal';
import ActionButtons from '../commons/ActionButtons';
import '../../assets/styles/editfeedback.scss';

/**
 * Edit Feedback Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
class EditFeedback extends Component {
  state = {
    question: this.props.question,
    type: this.props.type,
    startDate: this.props.startDate,
    duration: this.props.duration,
    closeModal: false,
  };

  /**
   * It closes a modal
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * It changes the state of the modal
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * Ensures that the state is updated basing
   * on the changes in the input fields
   *
   * @param {Object} target
   *
   * @returns {void}
   */
  handleInputChange = ({ target: { name, value } }) => {
    this.setState({ [name]: value });
  };

  /**
   * Handles editing feedback
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleEditFeedback = (event) => {
    event.preventDefault();
    /** submit logic here */
    this.handleCloseModal();
  };

  render() {
    const {
      question, type, startDate, duration, closeModal,
    } = this.state;

    return (
      <MrmModal
        title="Edit Question"
        buttonText="Edit"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="edit-feedback-modal"
        modalButtonClassName="edit-button"
      >
        <form className="modal-form" onSubmit={this.handleEditFeedback}>
          <Input
            labelName="Question"
            labelClass="label1"
            inputClass="input1"
            name="question"
            id="question"
            value={question}
            onChange={this.handleInputChange}
          />
          <Select
            labelText="Type"
            name="type"
            id="type"
            value={type}
            onChange={this.handleInputChange}
            options={[]}
            wrapperClassName="label1"
            placeholder="Rate"
            selectInputClassName="edit-feedback-select"
          />
          <Input
            labelName="Start Date"
            labelClass="label1"
            inputClass="input1"
            name="startDate"
            id="startDate"
            value={startDate}
            placeholder="03 Nov 2018"
            onChange={this.handleInputChange}
          />
          <Input
            labelName="Duration"
            labelClass="label1"
            inputClass="input1"
            name="duration"
            id="duration"
            value={duration}
            placeholder="3 Weeks"
            onChange={this.handleInputChange}
          />
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            actionButtonText="Save Changes"
          />
        </form>
      </MrmModal>
    );
  }
}

EditFeedback.propTypes = {
  question: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  startDate: PropTypes.string.isRequired,
  duration: PropTypes.string.isRequired,
};

export default EditFeedback;
