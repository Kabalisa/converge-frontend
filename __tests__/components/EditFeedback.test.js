import React from 'react';
import { shallow } from 'enzyme';
import EditFeedback from '../../src/components/EditFeedback';

describe('EditFeedback Component', () => {
  const props = {
    question: 'How would you rate the cleanliness of this room',
    type: 'rate',
    startDate: '3 Nov 2018',
    duration: '3 Weeks',
  };

  const wrapper = shallow(<EditFeedback {...props} />);

  const event = {
    preventDefault: jest.fn(),
    target: {
      name: '',
      value: '',
    },
  };

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a form', () => {
    const modalForm = wrapper.find('form');
    expect(modalForm).toHaveLength(1);
  });

  it('should handleCloseModal()', () => {
    wrapper.instance().handleCloseModal();
    expect(wrapper.state('closeModal')).toEqual(true);
  });

  it('should handleModalStateChange()', () => {
    wrapper.instance().handleModalStateChange();
    expect(wrapper.state('closeModal')).toBe(false);
  });

  it('should count the number of Select and Input element', () => {
    expect(wrapper.find('Input').length).toBe(3);
    expect(wrapper.find('SelectInput').length).toBe(1);
  });

  it('should respond to a state change', () => {
    wrapper.find('#question').simulate('change', {
      target: { name: 'question', value: 'Is there any missing meeting tools' },
    });
    expect(wrapper.find('#question').props().value).toBe('Is there any missing meeting tools');
    expect(wrapper.state('question')).toEqual('Is there any missing meeting tools');
  });

  it('should handleEditFeedback()', () => {
    wrapper.instance().handleEditFeedback(event);
    expect(wrapper.state('closeModal')).toEqual(true);
  });
});

