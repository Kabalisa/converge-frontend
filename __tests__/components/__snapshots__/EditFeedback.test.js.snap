// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditFeedback Component renders correctly 1`] = `
<MrmModal
  buttonText="Edit"
  className="edit-feedback-modal"
  closeModal={false}
  handleCloseRequest={[Function]}
  modalButtonClassName="edit-button"
  title="Edit Question"
>
  <form
    className="modal-form"
    onSubmit={[Function]}
  >
    <Input
      controlsClass=""
      id="question"
      inputClass="input1"
      labelClass="label1"
      labelName="Question"
      name="question"
      onChange={[Function]}
      placeholder=""
      type="text"
      value="How would you rate the cleanliness of this room"
    />
    <SelectInput
      id="type"
      isValue={false}
      labelText="Type"
      name="type"
      onChange={[Function]}
      options={Array []}
      placeholder="Rate"
      placeholderValue=""
      required={false}
      selectInputClassName="edit-feedback-select"
      value="rate"
      wrapperClassName="label1"
    />
    <Input
      controlsClass=""
      id="startDate"
      inputClass="input1"
      labelClass="label1"
      labelName="Start Date"
      name="startDate"
      onChange={[Function]}
      placeholder="03 Nov 2018"
      type="text"
      value="3 Nov 2018"
    />
    <Input
      controlsClass=""
      id="duration"
      inputClass="input1"
      labelClass="label1"
      labelName="Duration"
      name="duration"
      onChange={[Function]}
      placeholder="3 Weeks"
      type="text"
      value="3 Weeks"
    />
    <ActionButtons
      actionButtonText="Save Changes"
      cancelButtonText="CANCEL"
      isLoading={false}
      onClickCancel={[Function]}
      onClickSubmit={[Function]}
      withCancel={true}
    />
  </form>
</MrmModal>
`;
